examples-draft/SimpleExampleWF_Max.tla

```diff
@@ -0,0 +1,213 @@
+(*
+Helper invariant: mIsMax is stable.
+*)
+LEMMA mSpecStableMax == m!Spec => [](mIsMax => []mIsMax)
+  <1>1. m!Init /\ mIsMax => mIsMax
+        OBVIOUS
+  <1>2. mIsMax /\ [m!Next]_mX => mIsMax'
+        BY m!Assms DEF Max, mIsMax, m!Next
+  <1>q. QED BY <1>1, <1>2, PTL DEF m!Spec
+
+(*
+Helper invariant: Either X = Null XOR X = Max(A, B)
+*)
+mXInv ==
+  /\ (mX = m!Null) # mIsMax
+  /\ (mX = m!Null) \in BOOLEAN
+  /\ mIsMax \in BOOLEAN
+LEMMA mSpecXInv == m!Spec => []mXInv
+  <1>0. (mX # m!Null) \in BOOLEAN OBVIOUS
+  <1>x. mIsMax \in BOOLEAN BY DEF mIsMax
+  <1>1. m!Init => mXInv
+        <2>0. SUFFICES ASSUME m!Init PROVE mXInv OBVIOUS
+        <2>1. mX = m!Null BY <2>0 DEF m!Init
+        <2>2. ~mIsMax
+              BY <2>1, m!Assms, NoSetContainsEverything, MaxIsGEQ
+              DEF mIsMax, MaxProps, m!Null  
+        <2>q. QED BY <2>1, <2>2, <1>0, <1>x DEF mXInv
+  <1>2. mXInv /\ [m!Next]_mX => mXInv'
+        <2>1. SUFFICES ASSUME mXInv, m!Next PROVE mXInv'
+              BY DEF m!Next, mXInv, mIsMax, Max
+        <2>2. CASE mX = m!Null /\ ~mIsMax 
+              BY <2>1, <2>2, m!Assms
+              DEF mIsMax, Max, m!Next, mXInv
+        <2>3. CASE mX # m!Null /\ mIsMax
+              BY <2>1, <2>3, m!Assms
+              DEF mIsMax, Max, m!Next, mXInv
+        <2>4. QED BY <2>1, <2>2, <2>3, <1>0, <1>x, m!Assms DEF mXInv
+  <1>q. QED BY <1>1, <1>2, PTL DEF m!Spec
+
+(*
+Prove the that MaxAsFSM produces the same results as Max eventually.
+*)
+LEMMA mSpecEventuallyMax == m!Spec => <>mIsMax
+  <1>1. mA \in Nat /\ mB \in Nat /\ m!Null \notin Nat
+        BY m!Assms, NoSetContainsEverything DEF m!Null
+  (*
+  First in <1>2 we prove that either we have our target <>mIsMax or an
+  action making it true is enabled. That's an inductive invariant. Then
+  in <1>3 we show, that the action Next actually makes the predicate
+  mIsMax true. From this it follows that mIsMax will eventually be
+  reached.
+  *)
+  <1>2. m!Spec => [](<>mIsMax \/ <><<m!Next>>_mX)
+        <2> DEFINE P == mIsMax \/ ENABLED <<m!Next>>_mX
+        <2> HIDE DEF P
+        (*
+        In this case we cannot prove P from the definitions of mIsMax
+        and m!Next (as opposite to the ..._Flag example). We have to
+        assume m!Spec, and show P is true as an invariant (by induction in <2>1).
+
+        Then we use that in <2>2 to replace <>~ENABLED <<m!Next>>_x with mIsMax in:  
+        WF_mX(m!Next) = []( ([] ENABLED <<m!Next>>_x) => <><<m!Next>>_x)
+                      = [](~([] ENABLED <<m!Next>>_x) \/ <><<m!Next>>_x) 
+                      = []( (<>~ENABLED <<m!Next>>_x) \/ <><<m!Next>>_x)
+                      = []( (<>mIsMax               ) \/ <><<m!Next>>_x)
+
+        From this and the fact that Spec implies WF, we conclude <1>2.
+        *)
+        <2>1. m!Spec => []P
+              <3>1. m!Init => P
+                    <4>1. SUFFICES ASSUME m!Init PROVE P OBVIOUS
+                    <4>2. SUFFICES ASSUME TRUE PROVE ENABLED <<m!Next>>_mX BY DEF P
```

Comment:
These two steps could obviously be combined into

<4>1. SUFFICES ASSUME m!Init PROVE ENABLED <<m!Next>>_mX  BY DEF P